Cxspa 1638 ticket detailing failure path (new)

sushant-stha opened the pull request.

khristian23 committed:
Adding Customer Ticketing library (#15753)

* Customer Ticketing library creation

* Update README.md

* Additional changes

* Fix building errors and add feature-flag option

* Remove feature-flag and duplicated code

* Yarn config update

* Fix unit test runner

* Add schematics build

* Fix build

* Fix schematics reference

* Fix library configuration errors

* Fix customer ticketing selector prefix

* Fix prettier issues

* trigger build

* Configure Schematics

* Add customer ticketing library

* Fix typo

* Fix path

* Add index_spec.ts

* Ran 'yarn config:update'

* Test build

* Fix build

* Test build

* Review comments - part 1

* ran yarn cofig update

* Make structure consistent with other libs

* Prettier fix

* Try to fix build

* Fix unit test

* Fix integration test

* Fix typo

* Revert changes

* Remove unwanted file

* Add missing file

* new schematics changes

* Fix unit test

* Fix review comments

* Add cms components

* Add router dependency

* Updating root module

* removed dependency on itself

* merged latest develop to > epic > skeleton branch + revert lock changes

* finalize checking skeleton changes to be applied + removed extra break lines from tsconfigs

Co-authored-by: manpreetlamba4 <105296850+manpreetlamba4@users.noreply.github.com>
Co-authored-by: Manpreet Kaur <manpreet.kaur03@sap.com>
Co-authored-by: i540872 <sushant.shrestha@sap.com>
Co-authored-by: Brian Gamboc-Javiniar <brian_javiniar@hotmail.com>

manpreetlamba4 committed:
Merge branch 'develop' into epic/customer-ticketing

Parthlakhani committed:
refactor: update CMS component names (#15893)

* refactor: update CMS component names

* update component name

* update cypress.json

bgambocjaviniar committed:
updated env-cmdrc - need to revert before merging to develop - do not do what I just did

Parthlakhani committed:
feat: create skeleton for customer ticket details (#15953)

* feat: create skeleton for customer ticket details

* fix prettier

* remove hard code

* update endpoint according to backend

* remove unused module

Parthlakhani committed:
Merge branch 'develop' into epic/customer-ticketing

Parthlakhani committed:
(CXEC-14761) feat: add UI for customer ticket details (#16005)

* (CXEC-14761) feat: add UI for customer ticket details

* fix prettier issue

* correct spelling mistake

* update card text to adapt custom color

* Update card component and refactor customer ticketing details

* fix unit tests

* fix errors

* create enum for color and status

* update card component to use class instead of style

* extend custom class to truncated text

Parthlakhani committed:
Merge branch 'develop' into epic/customer-ticketing

adolfobee committed:
[Spartacus] - feature/CXEC-9977 - Adding initial changes to ticketing list (#16000)

* Adding initial changes to ticketing list

* Adding async, removing debug element

* Getting rid of ticketing list header

* Cleaning up ticketing list

* Some more cleaning

* prettify files

* update dependencies

Co-authored-by: Adolfo <adolfo.pereira@sap.com>
Co-authored-by: Brian Gamboc-Javiniar <brian_javiniar@hotmail.com>

Parthlakhani committed:
CXEC-15848: UI for reopen/close request (#16044)

* CXEC-15848: Add UI for reopen/close request
CMS component for reopen request button
CMS component for close request button
Reopen request dialog box with file upload and message

* remove console

* fix unit tests and prettier

* add default validations and validation checks

* add missing transaltion

* Add validations
add a method to file form validators to check the maximum file entries
add validations to customer ticketing file upload

* Update feature-libs/customer-ticketing/components/customer-ticketing.service.ts

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

* Add file upload notes

* fix lint errors

* refactor config and file upload
moved default config to root
added template reference to fix file upload button issue in import products

* fix click issue on upload file

* refactor object name

* fix prettier issue

* add todo

* add placeholders selectors

* update css

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

Parthlakhani committed:
Merge branch 'develop' into epic/customer-ticketing

Parthlakhani committed:
CXEC-14759: Integrate api for ticket details (#16056)

* CXEC-14759: Integrate getTicket API

* refactor logic for preconditions

* Add unit tests

* add unit tests

* unit tests for component

* prettier fix

* refactor code

* update unit test

* expose ticket state for spinner

* add method name to facade

* Update feature-libs/customer-ticketing/components/customer-ticketing-details/customer-ticketing-details.component.spec.ts

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

* update unit tests and model

* fix unit test and remove comments

* remove comments

* change model

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

Parthlakhani committed:
Create abstract dialog component for customer ticketing (#16113)

* Create abstract dialog component

* add date format

* refactor code

manpreetlamba4 committed:
Add initial changes for create ticket feature (#16063)

* Add initial changes for create ticket feature

* Correct css for Add button

* Adopt abstract component for dialog

* remove export

* Use distinct selector for create and reopen dialog

* update selector name and usage of classes

adolfobee committed:
Adding ticketing list, styles and table - Feature/CXEC-15925 (#16059)

* Adding translations needed

* Creating table html

* Creating custom scss for table

* Importing the new style

* Adding tickets and data to the ticketing list

* Adding ticket counts and select dropdown

* Adding data to the select

* Styling select dropdown

* Adding missing translations

* Adding styling in the components folder

* Adding missing translations for mobile view

* Adding styles for mobile

* Removing import

* Importing icons

* Bringing icons to the component

* Adding table for tablet and mobile view

* Adding correct background color to select

* Modifying dates labels for mobile view

* Adding formatting to mobile view

* Removing unneeded options

* Adding translations object

* Putting all translations in their own object

* Adjusting translations in html

* Getting rid of unused container

* Using sass placeholder selector

* Removing unused class

* Removing modified table styles

* Modifying customer ticketing styling

* Removing arrow icon

* Adding testing module

* Adding style changes to table

* More table styling

* Removing comments

* Implementing onInit and adding flag

* Adding translation

* Adding sorting and pagination

* Adding styles

* Importing missing modules

* Adding models to ticketing list

* Adding color to ticket status

* Adding status color based on ID

* Adding styles for colors

* Adding routing module to ticket list

* Adding padding to first element of table

* Getting rid of the list view in mobile

* Adding style for mobile table

* Lint

* Fixing styling

* Adding description to table

* Adding tab indexes

* Adding in process status and modifying color

* Fixing tests

* Removing unnecessary information

* Linting styles

* Closing container

* Linting

* Renaming Constants adding enum

* Renaming variable, reducing sort func

* Reworking sorting and page change

* Addressing pr comments

* Adding routing service to ticketing list component

* Modifying styling

* Showing title when no items on the list

* Passing ticketCode object

* Hiding sorting when list is empty

* Moving container

Co-authored-by: Adolfo <adolfo.pereira@sap.com>

manpreetlamba4 committed:
Feature/cxec 16778 : Add Categories and Associated Objects API in Dialog (#16301)

* Integrate APIs in dialog

* Correct typo

* Fix review comments

* Address Review comments

* Fix review comments

* Adopt review comments

Parthlakhani committed:
CXEC-17407: Close/reopen requests (#16128)

* Add close request dialog box

* integrate close and reopen ticket apis

* add spinner while data is loading

* fix unit tests

* remove button type and fix constant name mistake

* fix validation checks

* add unit tests

* refactor code

* eliminate use of details service

* remove serializer

* add unit tests

* add unit tests

* fix variable name

* fix unit tests

Parthlakhani committed:
CXEC-14762: UI for messaging and file upload (#16085)

* create shared component for messaging
shared messaging and avatar component
customer ticketing messaging component
validation for file upload

* update css

* fix unit test

* update css

* fix linter errors

* fix prettier

* add check for addedByAgent

* refactor code

* add unit tests

* add unit tests for messaging

* update constant name

* Update projects/storefrontlib/shared/components/chat-messaging/avatar/avatar.component.ts

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

* add dateFormat constant

* fix tab index for file upload

* fix tabbing and area label for messaging

* fix AMP bugs

* remove bootstarp class

* add label to fix AMP bug

* fix css issue

* add character limit alert for screen readers

* fix unit tests

* change label name

* remove button type

* refactor code
fix imports
remove scroll logic
add unit test

* refactor css

* fix unit tests

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

ishrarhussain committed:
CXEC-16052: Cypress e2e test for viewing ticket details (#16368)

* CXEC-16052: Cypress e2e test for viewing ticket details

This adds the cypress e2e test in customer ticketing for the happy path of viewing the ticket details of an already created ticket.

Issue #CXEC-16052 (https://jira.tools.sap/browse/CXEC-16052)

+ Minor changes as per PR review

+ Declare constants for indices of UI options

Parthlakhani committed:
cxec-14760: add new message and upload/download attachment

* create shared component for messaging
shared messaging and avatar component
customer ticketing messaging component
validation for file upload

* update css

* fix unit test

* update css

* fix linter errors

* fix prettier

* add check for addedByAgent

* refactor code

* add unit tests

* add unit tests for messaging

* update constant name

* Update projects/storefrontlib/shared/components/chat-messaging/avatar/avatar.component.ts

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

* add dateFormat constant

* fix tab index for file upload

* fix tabbing and area label for messaging

* fix AMP bugs

* remove bootstarp class

* add label to fix AMP bug

* fix css issue

* add character limit alert for screen readers

* fix unit tests

* Add close request dialog box

* integrate close and reopen ticket apis

* add spinner while data is loading

* fix unit tests

* change label name

* remove button type and fix constant name mistake

* remove button type

* fix validation checks

* add unit tests

* refactor code

* eliminate use of details service

* remove serializer

* refactor code
fix imports
remove scroll logic
add unit test

* refactor css

* add unit tests

* add unit tests

* integrate download attachment api

* fix errors

* download attachment api integration

* upload attachment while reopening

* Refactor variable name

* move reverse logic to adapter

* fix console error

* fix unit test

* fix unit tests and errors

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

adolfobee committed:
Retrieving data from OCC endpoints for ticketing list - Feature/CXEC-15923 (#16130)

* Adding translations needed

* Creating table html

* Creating custom scss for table

* Importing the new style

* Adding tickets and data to the ticketing list

* Adding ticket counts and select dropdown

* Adding data to the select

* Styling select dropdown

* Adding missing translations

* Adding styling in the components folder

* Adding missing translations for mobile view

* Adding styles for mobile

* Removing import

* Importing icons

* Bringing icons to the component

* Adding table for tablet and mobile view

* Adding correct background color to select

* Modifying dates labels for mobile view

* Adding formatting to mobile view

* Removing unneeded options

* Adding translations object

* Putting all translations in their own object

* Adjusting translations in html

* Getting rid of unused container

* Using sass placeholder selector

* Removing unused class

* Removing modified table styles

* Adding query to adapter and connectors

* Adding new queries to services

* Adding service tests

* Modifying OCC adapter

* Adding tests

* Adding new query to occ endpoints and config

* Modifying customer ticketing model

* Adding query to facade and events

* Adding ticket list to converter

* Modifying customer ticketing styling

* temp

* Removing header when no tickets

* Grabbing ticket list

* Getting whole customer ID

* Getting name for category and status on ticket list

* Adding color to ticket status lists

* Changing sort labels

* Fixing console errors when initiating component

* Adding fetch list when sorting and pagination

* Adding pagination and sorting

* Adding correct total results to tickets count

* Removing providers for create ticket

* Modifying style for create component

* Importing create ticket into list

* list service tests

* Fixing adapter tests

* Adding unit tests

* Testing empty ticket list

* Addressing comments

* Reload events

* Event to login logout

* Removing commented code

* Removing event from listener

* Fixing service tests

* Adding routing service to unit test

* Fixing tests

* Addressing comments

Co-authored-by: Adolfo <adolfo.pereira@sap.com>

osahtout committed:
CXEC-15125 implement create ticket e2e (#16375)

* CXEC-15125 implement create ticket e2e

Addressed comments regarding wait

Removed magic numbers

Many improvements

rebase changes

* implement own wait for page

* fix apparel link

* comments

t#

Co-authored-by: Omar Sahtout <omar.sahtout@sap.com>

manpreetlamba4 committed:
Feature/cxec 15123 Integrate API to create a ticket (#16315)

* Integrate APIs in dialog

* Correct typo

* Fix review comments

* Address Review comments

* Fix review comments

* Adopt review comments

* API to create ticket

* Add missing code

* Fix review comments

* Update feature-libs/customer-ticketing/assets/translations/en/customer-ticketing.ts

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

* Correct spelling

* Update text

* Update text

* Fix compiler error

* Group create ticket texts

* Remove use of subscribe

* Fix review comments

* Remove null for fileupload

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

adolfobee committed:
Deleting pagination at the top, adding create ticket on empty list (#16407)

Co-authored-by: Adolfo <adolfo.pereira@sap.com>

sushant-stha committed:
Adding failure path for Ticket Detailing

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
   visitPage('apparel-uk-spa/en/GBP/my-account/support-tickets', 'apparelTicketListingPage');
 }
+
+export function getTicketDetailsFromTicketListView(){
```
Other functions use TicketListingPage, please adhere to the convention :)

sushant-stha replied:
Done

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
 }
+
+export function getTicketDetailsFromTicketListView(){
+  let ticketDetails = [];
+  const row = cy.get('cx-customer-ticketing-list').find('tbody').get('tr').eq(1);
+  row.get('td').eq(0).invoke('text').then((x) => ticketDetails.push(x));
+  row.get('td').eq(1).invoke('text').then((x) => ticketDetails.push(x));
+  row.get('td').eq(5).invoke('text').then((x) => ticketDetails.push(x));
+
+   return ticketDetails;
```
Nice idea, some remarks:
1) Avoid using magic numbers 0 1 5 have a look at: https://github.com/SAP/spartacus/blob/91a7e38c0cea1e7e46d453bd3c4db7ce952abe26/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts#L123

2) There is an interface hosting ticket details, please use it: 
https://github.com/SAP/spartacus/blob/91a7e38c0cea1e7e46d453bd3c4db7ce952abe26/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts#L20

3) Add return type to the function signature

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+   return ticketDetails;
+}
+
+export function verifyTicketDetails(ticketDetails){
+  //assert title when available
+  cy.get(':nth-child(1) > cx-card > .cx-card > .card-body > .cx-card-container > .cx-card-label-container > :nth-child(1) > .cx-card-label').contains(ticketDetails[1]);
+  cy.get(':nth-child(4) > cx-card > .cx-card > .card-body > .cx-card-container > .cx-card-label-container > :nth-child(1) > .cx-card-label').contains(ticketDetails[2]);
+}
```
1.  Please specify parameter Type as part of the signature
2. Avoid using magic number so we can tell what is the 1st and the 4th child just by reading the code
3. Modify the name of the function to better describe its purpose, you might need to add the page expect to be in

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get(':nth-child(4) > cx-card > .cx-card > .card-body > .cx-card-container > .cx-card-label-container > :nth-child(1) > .cx-card-label').contains(ticketDetails[2]);
+}
+
+export function createNewTicket(){
+  const testTicketDetails: TestTicketDetails = {
+    subject: 'Entering a subject',
+    message: 'Typing a message',
+    category: TestCategory.complaint,
+  };
+
+  loginRegisteredUser();
+  visitElectronicSupportTicketPage();
+  openCreateTicketModal();
+  fillTicketDetails(testTicketDetails);
+  clickSubmitAndVerifyRequestCompleted();
+  verifyCreatedTicketDetails(testTicketDetails);
+}
```
This is exactly what I mentioned in your other PR:
https://github.com/SAP/spartacus/pull/16418#discussion_r1013435928

But I will prefer you parameterize the TestTicketDetails so its usage is more flexible
In addition, please add `Journey` to your function name so its purpose its clear

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+
+export function verifyTicketDetails(ticketDetails){
+  //assert title when available
+  cy.get(':nth-child(1) > cx-card > .cx-card > .card-body > .cx-card-container > .cx-card-label-container > :nth-child(1) > .cx-card-label').contains(ticketDetails[1]);
```
Do you really need to specify all the path of wrapping DOM elements?
could it be just enough to do something like:
cy.get(':nth-child(FIRST_SOMETHING)  .cx-card .cx-card-label-container  :nth-child(1) .cx-card-label')

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+}
+
+export function clickCloseRequestButton(){
+  cy.get(':nth-child(4) > cx-card > .cx-card > .card-body > .cx-card-container > .cx-card-label-container > :nth-child(1) > .cx-card-label').contains("Open");
```
Please replace `4` by proper constant
Try to reduce the path of DOM elements being parsed

osahtout replied:
I think this should work 
```
export function clickCloseRequestButton() {
  cy.get('cx-customer-ticketing-close').click();
}
```

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+
+export function clickCloseRequestButton(){
+  cy.get(':nth-child(4) > cx-card > .cx-card > .card-body > .cx-card-container > .cx-card-label-container > :nth-child(1) > .cx-card-label').contains("Open");
+  cy.get('.btn.btn-block.btn-action"').click();
```
can you use an existent function like `clickCancel`?

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get(':nth-child(4) > cx-card > .cx-card > .card-body > .cx-card-container > .cx-card-label-container > :nth-child(1) > .cx-card-label').contains("Open");
+  cy.get('.btn.btn-block.btn-action"').click();
+  cy.get('.cx-customer-ticket-form-body').should('exist');
+  cy.get('label > .form-control').last().type("Closing Ticket");
+  cy.get('.cx-customer-ticket-form-footer > .btn-primary').click();
```
Hard to know what is the purpose of these lines without reading the DOM generated
Could you use aria labels of UI element roles instead?

khristian23 replied:
This function is no longer needed.
Please reuse the following:
https://github.com/SAP/spartacus/blob/679d920a73a8cb7fef2a769e065e2f39713249f0/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-details.ts#L19

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get('.cx-customer-ticket-form-footer > .btn-primary').click();
+}
+
+export function verifyStatusOnListingPage(status){
+  const row = cy.get('cx-customer-ticketing-list').find('tbody').get('tr').eq(1);
+  row.get('td').eq(5).contains(status);
+}
```
No magic numbers, please.
Make the function name more concrete. Maybe something like verifyStatusOfTicketInPositionAtTicketListingPage(status: string, ticketPosition: number)
Do not restrict yourself from adding long names to JS functions.

osahtout replied:
I think this can work 
```export function verifyClosedTicketIsStillInTicketListing() {
  const row = cy.get('cx-customer-ticketing-list').get('tbody').get('tr').eq(FIRST_ROW_TICKET_LIST);
  row.get('td').eq(STATUS_COLUMN).contains("Closed");
}
```
You can add a parameter if you need as well

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  row.get('td').eq(5).contains(status);
+}
+
+export function verifyMessageBoxExists(doesExist){
+  if(doesExist){
+    cy.get('.form-control').should('exist');
+  } else {
+    cy.get('.form-control').should('not.exist');
+  }
+}
```
Do not use booleans in functions to diversify the implementation.
In this case you need to write two methods:
https://github.com/ryanmcdermott/clean-code-javascript#dont-use-flags-as-function-parameters

khristian23 replied:
What is the purpose of this function?
I cannot tell just by looking at its implementation since. ".form-control" is used in multiple places

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  }
+}
+
+export function clickReopenRequestButton(){
```
SAme comments as in `clickCloseRequestButton` above
Is it possible to pass a variable with the button Id or name and create a generic function like:
clickOnButtonInPopup(buttonName: string)

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get('.cx-customer-ticket-form-footer > .btn-primary').click();
+}
+
+export function verifyMessagesAreBeingPopulated(){
```
What kind of message?

sushant-stha replied:
The interaction between customer and agent. But there since there is a chance the customer has just created ticket which means only message will be there so i just created a assert on customers side of the tab

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get('.cx-message-left-align-text').should('exist');
+}
+
+export function addFilesInChatBox(){
```
Maybe you can reuse:
https://github.com/SAP/spartacus/blob/91a7e38c0cea1e7e46d453bd3c4db7ce952abe26/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts#L89

sushant-stha replied:
This one wont be possible as the addFIle from Omar is targeting the popup for create new ticket and mine is for the upload button on the footer of ticket detailing

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get('p').contains("test.docx");
+}
+
+export function sendMessageAsUser(){
```
I don't understand the purpose of the function, please rename

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+}
+
+export function sendMessageAsUser(){
+  cy.get('.form-control').type("Update");
```
MAybe extract to function like verifyFormXYZIsEditable

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +149,77 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+
+export function sendMessageAsUser(){
+  cy.get('.form-control').type("Update");
+  cy.get('.cx-message-input > .btn').click();
```
Try to reuse existing functions.
If it does not exist, please create a similar function to `clickCancel` and `clickSubmit`

sushant-stha replied:
Created new function clickSend()

manpreetlamba4 committed:
Feature/cxec 15124 Integrate API to Upload Attachment (#16383)

* Integrate APIs in dialog

* Correct typo

* Fix review comments

* Address Review comments

* Fix review comments

* Adopt review comments

* API to create ticket

* Add missing code

* Fix review comments

* Update feature-libs/customer-ticketing/assets/translations/en/customer-ticketing.ts

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

* Correct spelling

* Add attachment api for create ticket

* Add method in facade

* Fix compile error

* Remove duplocate code from merge

* Remove multiple subscribe

* Fix review comments

* Change correct response type

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

Parthlakhani committed:
CXEC-20553: fix file upload restrictions (#16408)

sushant-stha committed:
Adding refactoring as PR comments

Parthlakhani committed:
fix max entries issue (#16428)

sushant-stha committed:
Refactoring as per PR comments

sushant-stha committed:
Merge branch 'epic/customer-ticketing' of https://github.com/SAP/spartacus into CXSPA-1638-ticket-detailing-failure-path-new

sushant-stha committed:
Adding refactorign as per PR review

adolfobee committed:
Addressing accesibility pointers (#16427)

Co-authored-by: Adolfo <adolfo.pereira@sap.com>

ishrarhussain committed:
CXEC-17550: Navigate to Ticket Listing in case of 404 Not Found Ticket Detail (#16405)

* CXEC-17550: Navigate to Ticket Listing in case of 404 Not Found Ticket Detail

Jira Issue # CXEC-17550 (https://jira.tools.sap/browse/CXEC-17550)

* Simplify error details retrieval with constants

+ Added separate module for constants in error details strings

Co-Authored-By: Christian Montoya <christian.montoya.b@gmail.com>

* Added check on http status code

* Added a new http-interceptor as per request on the PR review

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

sushant-stha committed:
Merge branch 'epic/customer-ticketing' of https://github.com/SAP/spartacus into CXSPA-1638-ticket-detailing-failure-path-new

sushant-stha committed:
Refactoring as per PR comments

sushant-stha committed:
Adding mobile viewport back

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -21,17 +27,27 @@ export interface TestTicketDetails {
+  cy.reload();
+}
+
+// export function visitHomePage(){
+//   cy.visit('/');
+//   cy.get('#cx-header', { timeout: 10000 }).should('be.visible');
+// }
+
```
Please remove

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -21,17 +27,27 @@ export interface TestTicketDetails {
   login();
 }
 
+export function loginAsLindaWolf() {
+  cy.requireLoggedIn(myCompanyAdminUser);
+  cy.reload();
+}
```
I guess this is no longer needed

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -52,6 +68,10 @@ export function clickFirstTicketFromTicketListing(){
   cy.get(`#ticketing-list-table tbody tr:nth-child(${FIRST_TICKET_ROW_INDEX}) .cx-ticketing-list-data:nth-child(${FIRST_TICKET_COLUMN_INDEX}) a.cx-ticketing-list-value`).click();
 }
 
+export function clickSecondTicketFromTicketListing(){
+  cy.get(`#ticketing-list-table tbody tr:nth-child(${SECOND_TICKET_ROW_INDEX}) .cx-ticketing-list-data:nth-child(${FIRST_TICKET_COLUMN_INDEX}) a.cx-ticketing-list-value`).click();
+}
+
```
Your branch needs to be rebased.  Omar did some refactoring
For instance, you no longer need to create a new function for clicking the Second, Third, and Fourth items.
Now you can call. "clickTicketInRow(SECOND_ROW)" in your journey

https://github.com/SAP/spartacus/blob/679d920a73a8cb7fef2a769e065e2f39713249f0/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts#L22

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
 }
+
+export function extractTicketDetailsFromFirstRowInTicketListingPage(): TestTicketDetails {
+  let testTicketDetails: TestTicketDetails = {
+    subject: "Temp",
+    message: "Temp",
+    category: TestCategory.complaint,
+    id: "000000",
+    status: "Open"
+  };
```
No need to make it LET because its reference is not mutating, what is mutating are its internal values. So CONST should work.
Also, you can create an empty TestTicketDetails object and fill it with just the content you find.

```suggestion
  const testTicketDetails: TestTicketDetails;
```

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+    status: "Open"
+  };
+
+  const row = cy.get('cx-customer-ticketing-list').find('tbody').get('tr').eq(FIRST_ROW);
+  row.get('td').eq(ID_COLUMN).invoke('text').then((x) => {
+    x = x.substring(ID_DELIMITER);
+    testTicketDetails.id = x.toString();
+  });
+  row.get('td').eq(SUBJECT_COLUMN).invoke('text').then((x) => {
+    x = x.substring(SUBJECT_DELIMITER);
+    testTicketDetails.subject = x.toString();
+  });
+  row.get('td').eq(STATUS_COLUMN).invoke('text').then((x) => {
+    x = x.substring(STATUS_DELIMITER);
+    testTicketDetails.status = x.toString();
+  });
```
```suggestion
  cy.get('cx-customer-ticketing-list').find('tbody').get('tr').eq(FIRST_ROW).within(row => {
     row.get('td').eq(ID_COLUMN).invoke('text').then(id => {
        testTicketDetails.id = id.substring(ID_DELIMITER);
      });
      row.get('td').eq(SUBJECT_COLUMN).invoke('text').then(subject => {
         testTicketDetails.subject = subject.substring(SUBJECT_DELIMITER);
      });
      row.get('td').eq(STATUS_COLUMN).invoke('text').then((x) => {
         testTicketDetails.status = x.substring(STATUS_DELIMITER);
      });
  });
```

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+}
+
+export function verifyTicketDetailsByComparingTicketHeaderToExtractedDetails(ticketDetails: TestTicketDetails){
+  //assert title when available
```
Please remove

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+export function verifyTicketDetailsByComparingTicketHeaderToExtractedDetails(ticketDetails: TestTicketDetails){
+  //assert title when available
+
+  cy.get('.cx-card-label').eq(0).then((x) => cy.wrap(x).should('include.text', ticketDetails.id));
+  cy.get('.cx-card-label').eq(3).then((x) => cy.wrap(x).should('include.text', ticketDetails.status));
```
Please replace the magic numbers (0,3) with constants
replace X with the proper variable name

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get('.cx-card-label').eq(3).then((x) => cy.wrap(x).should('include.text', ticketDetails.status));
+}
+
+export function createNewTicket(){
+  const testTicketDetails: TestTicketDetails = {
+    subject: 'Entering a subject',
+    message: 'Typing a message',
+    category: TestCategory.complaint,
+  };
+
+  visitElectronicTicketListingPage();
+  openCreateTicketPopup();
+  fillTicketDetails(testTicketDetails);
+  clickSubmit();
+  verifyCreatedTicketDetails(testTicketDetails);
+}
```
Please reuse existing function:
export function createTicket(ticketDetails: TestTicketDetails) {

https://github.com/SAP/spartacus/blob/679d920a73a8cb7fef2a769e065e2f39713249f0/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/create-tickets.ts#L77

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  verifyCreatedTicketDetails(testTicketDetails);
+}
+
+export function verifyTicketStatusIsOpenInTicketDetailsPage(){
+  cy.get('.cx-card-label').eq(3).contains("Open");
+}
+
+export function verifyTicketStatusIsClosedInTicketDetailsPage(){
+  cy.get('.cx-card-label').eq(3).contains("Closed");
+}
```
No longer needed.
Please reuse:
https://github.com/SAP/spartacus/blob/679d920a73a8cb7fef2a769e065e2f39713249f0/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts#L113

usage:
customerTicketing.verifyStatusOfTicketInList(FIRST_TICKET, TestStatus.Open)
customerTicketing.verifyStatusOfTicketInList(FIRST_TICKET, TestStatus.Closed)

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get('.cx-customer-ticket-form-footer > .btn-primary').click();
+}
+
+export function verifyTicketStatusIsOpenOnListingPage(){
+  const row = cy.get('cx-customer-ticketing-list').find('tbody').get('tr').eq(1);
+  row.get('td').eq(5).contains(" Open ");
+}
```
No longer needed, please reuse:

https://github.com/SAP/spartacus/blob/679d920a73a8cb7fef2a769e065e2f39713249f0/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts#L113

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  row.get('td').eq(5).contains(" Open ");
+}
+
+export function verifyTicketStatusIsClosedOnListingPage(){
+  const row = cy.get('cx-customer-ticketing-list').find('tbody').get('tr').eq(1);
+  row.get('td').eq(5).contains(" Closed ");
+}
```
No longer needed:

https://github.com/SAP/spartacus/blob/679d920a73a8cb7fef2a769e065e2f39713249f0/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts#L113

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  }
+}
+
+export function reopenTicket(){
+  cy.get('.btn.btn-block.btn-action"').click();
+  cy.get('.cx-customer-ticket-form-body').should('exist');
+  cy.get('label > .form-control').last().type("Reopening Ticket");
+  cy.get('.cx-customer-ticket-form-footer > .btn-primary').click();
+}
```
Could you please add a similar function for reopening a ticket like this:

https://github.com/SAP/spartacus/blob/679d920a73a8cb7fef2a769e065e2f39713249f0/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-details.ts#L19

Use the above function as a template

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get(".cx-message-left-align-text").eq(count_of_message_after_post-1).contains("Update ticket with comments");
+}
+
+export function navigateBackToPreviousPage(){
+  cy.go('back');
+}
```
Please move to customer Ticketing commons file

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  verifyLastMessageWasPosted();
+}
+
+export function clickSend(){
+  cy.get('.cx-send').click();
+}
+
+export function postMessageAsCustomerIntoChatBox(){
+  cy.get('.form-control').type("Update ticket with comments");
+}
```
Please reuse the following function, to replace the above two

https://github.com/SAP/spartacus/blob/679d920a73a8cb7fef2a769e065e2f39713249f0/projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-details.ts#L28

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get('.cx-customer-ticket-form-footer > .btn-primary').click();
+}
+
+export function verifyAllCustomerMessagesAreBeingPopulatedInChatHistory(){
+  verifyLastMessageWasPosted();
+}
```
I don't think this wrapper function is needed you can use verifyLastMessageWasPosted directly

khristian23 commented on projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing.ts:
```diff
@@ -149,3 +169,112 @@ export function verifyTicketDoesNotExist(ticketDetails: TestTicketDetails) {
+  cy.get('.form-control').type("Update ticket with comments");
+}
+
+export function verifyLastMessageWasPosted(){
+  let count_of_message_after_post = 2;
+  cy.get(".cx-message-left-align-text", {timeout: 100000}).should('have.length', count_of_message_after_post);
+  cy.get(".cx-message-left-align-text").eq(count_of_message_after_post-1).contains("Update ticket with comments");
+}
```
```suggestion
export function verifyLastMessageWasPosted(message: string){
  const position_of_last_message = cy.get('selector to get last message');
  cy.get(".cx-message-left-align-text").eq(position_of_last_message).contains(message);
}
```

khristian23 reviewed (changes requested):
Hey @sushant-stha there was a big refactoring in the last push of e2e coding.
Please rebase your changes.  Many functions that you need already exist in a more generic way.  I have pointed them out in the comments
